formats/protobuf/commonMain/src/kotlinx/serialization/protobuf/internal/ProtobufDecoding.kt

```diff
@@ -204,16 +245,28 @@ internal open class ProtobufDecoder(
         while (true) {
             val protoId = reader.readTag()
             if (protoId == -1) { // EOF
+                for (i in 0 until descriptor.elementsCount) {
```

Comment:
I'm afraid that this might fall into the 'accidentally quadratic' case. Have you tried to use a bit mask in Int as a replacement for FILLED? We still would need an array thought (because there are 3 possible states), but we wouldn't do a linear lookup every time, `Int.countLeadingZeroBits` or smth like that would be sufficient

Reply:
Another way to implement it are two arrays with bit values (final absence reading mark and null/empty collection mark)